### PR TITLE
Add AppDefaults 

### PR DESCRIPTION
**Pull request by jiaqiluo**
Issue: https://github.com/rancher/rancher/issues/36827

# Problem:
We use channelserver to host rke2/k3s version information in rancher/rancher, and we need the ability to set different default versions for different rancher versions.  
This ability should be general enough to support any app and not tied to Rancher. 

# Solution:
This PR adds the support for setting app defaults in the config file. 
An example of the value looks like this 
```
appDefaults:
    - appName: rancher
      defaults:
        - appVersion: '> 2.6.2-0 < 2.6.3-0'
          defaultVersion: '1.20.x'
        - appVersion: '> 2.6.2-0 < 2.6.4-0'
          defaultVersion: '1.21.x'
        - appVersion: '> 2.6.4-0'
          defaultVersion: '1.22.x'
    - appName: anotherApp
      defaults:
        - appVersion: '1.2.x'
          defaultVersion: '1.19.5'
```

# Problem:
In the case where the 404 NotFound is returned from the HTTP GET request, channelserver cannot parse the response's data and is stuck in returning errors. 

# Solution:
This PR adds checks for the response's status code and continues to parse the data only if it is 200.

# Problem:
The default value for the flag `path-prefix` is not applied. 

# Solution:
It is a bug in the urfave/cli module: https://github.com/urfave/cli/issues/1121
This PR bumps github.com/urfave/cli/v2 to v2.4.0 to get the fix 

# Others Improvements:
- update error messages to be clear 
- update the readme and default config file

### CODE REVIEW

**Review by paynejacob — APPROVED** [review]

File: pkg/config/get.go [hunk 1]
```diff
@@ -146,3 +149,37 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
...
+
+	if subKey == "" {
+		if err := yaml.Unmarshal(content, &allConfigs); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := yaml.Unmarshal(content, &data); err != nil {
+			return nil, err
+		}
+		data, _ = data[subKey].(map[string]interface{})
```
Comment (thedadams) [inline pkg/config/get.go]: Let's check this potential casting error, just in case.

File: pkg/config/get.go [hunk 2]
```diff
@@ -45,6 +45,9 @@ func get(ctx context.Context, url Source) ([]byte, error) {
 	if err != nil {
 		return nil, err
 	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("status %v", resp.Status)
+	}
```
Comment (thedadams) [inline pkg/config/get.go:52]: This needs to come after the `defer resp.Body.Close()`

File: main.go [hunk 3]
```diff
@@ -56,6 +57,12 @@ func main() {
 			EnvVars:     []string{"CHANNEL_SERVER_VERSION"},
 			Destination: &ChannelServerVersion,
 		},
+		&cli.StringFlag{
+			Name:        "app-name",
+			Usage:       "the app for which the app default versions are for",
```
Comment (thedadams) [inline main.go]:
```suggestion
			Usage:       "the app for which to retrieve default versions",
```

**Review by thedadams — CHANGES_REQUESTED** [review]
A few minor things.

**Comment by jiaqiluo** [comment]
@thedadams  Please check[ the latest commit](https://github.com/rancher/channelserver/pull/17/commits/51f595abf5f59ec0198384b5fb0b519028aa34a4) that addresses your comments.

File: pkg/config/get.go [hunk 4]
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
...
+		if err := yaml.Unmarshal(content, &allConfigs); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := yaml.Unmarshal(content, &data); err != nil {
+			return nil, err
+		}
+		data, ok := data[subKey].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("failed to do the type assertion")
```
Comment (thedadams) [inline pkg/config/get.go]:
```suggestion
			return nil, fmt.Errorf("content at key %s expected to be map[string]interface{}, found %T instead", subKey, data[subKey])
```

File: pkg/config/get.go [hunk 5]
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
...
+
+	if subKey == "" {
+		if err := yaml.Unmarshal(content, &allConfigs); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := yaml.Unmarshal(content, &data); err != nil {
+			return nil, err
+		}
+		data, ok := data[subKey].(map[string]interface{})
```
Comment (thedadams) [inline pkg/config/get.go]: Avoid variable shadowing so we can make the error message more descriptive.
```suggestion
		subData, ok := data[subKey].(map[string]interface{})
```

File: pkg/config/get.go [hunk 6]
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
...
+		if !ok {
+			return nil, fmt.Errorf("failed to do the type assertion")
+		}
+		if err := convert.ToObj(data, &allConfigs); err != nil {
```
Comment (thedadams) [inline pkg/config/get.go]:
```suggestion
		if err := convert.ToObj(subData, &allConfigs); err != nil {
```

**Review by thedadams — CHANGES_REQUESTED** [review]

**Comment by jiaqiluo** [comment]
@thedadams  Please check the latest commit: https://github.com/rancher/channelserver/pull/17/commits/41efe6ca59f80683dd0e006597a478d16de0ee51

File: pkg/config/get.go [hunk 7]
```diff
@@ -146,3 +150,40 @@ func GetGHReleases(ctx context.Context, client *github.Client, owner, repo strin
...
+		if err := yaml.Unmarshal(content, &allConfigs); err != nil {
+			return nil, err
+		}
+	} else {
+		if err := yaml.Unmarshal(content, &data); err != nil {
+			return nil, err
+		}
+		subData, ok := data[subKey].(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf("content at key %s expected to be map[string]interface{}, found %T instead", subKey, subData[subKey])
```
Comment (thedadams) [inline pkg/config/get.go]:
```suggestion
			return nil, fmt.Errorf("content at key %s expected to be map[string]interface{}, found %T instead", subKey, data[subKey])
```

**Review by thedadams — CHANGES_REQUESTED** [review]

**Review by thedadams — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**